### PR TITLE
gnrc: integrate gnrc_netif2

### PR DESCRIPTION
**Pull request by miri64**
Does not link ATM, since I did not bother to port the old NC and NDP (will provide a port for #7014 in a separate PR though [#7456]) and those modules now complain that they don't find the old interfaces functions. Since this change is unintrusive (and I do not provide special new applications for this), this should be okay.

~~Depends on #7410 and #7409 and all their dependencies.~~

This PR is part of the network layer remodelling effort:
PR dependencies

### CODE REVIEW

**Comment by miri64** [comment]
Rebased and adapted to current master and current dependencies

**Comment by miri64** [comment]
Rebased to current master and dependencies

**Comment by miri64** [comment]
Rebased to current master

**Comment by miri64** [comment]
Rebased to current master and current dependencies.

**Comment by miri64** [comment]
Rebased to current master, no longer depending on other PRs.

File: examples/gnrc_minimal/main.c [hunk 1]
```diff
@@ -22,16 +22,38 @@
...
+#ifdef MODULE_GNRC_NETIF2
+    /* get interface and print their addresses */
+    gnrc_netif2_t *netif = NULL;
+    while ((netif = gnrc_netif2_iter(netif))) {
+        ipv6_addr_t ipv6_addrs[GNRC_NETIF2_IPV6_ADDRS_NUMOF];
+        int res = gnrc_netapi_get(netif->pid, NETOPT_IPV6_ADDR, 0, ipv6_addrs,
+                                  sizeof(ipv6_addrs));
+
+        /* res / 16 (res / sizeof(ipv6_addr_t)) */
+        for (int i = 0; i < (res >> 4); i++) {
```
Comment (cgundogan) [inline examples/gnrc_minimal/main.c]: what's  the point in writing `(res >> 4)` when you need to document that?
Just use `(res / sizeof(ipv6_addr_t)`. Every compiler will certainly use a shift for that (:

**Comment by cgundogan** [comment]
Hm, doesn't look that *unintrusive* to me. A lot of modules are touched and I cannot test them all. Just by reviewing (i.e. reading code), the changes look reasonable .. I'd like a second opinion from @haukepetersen or @smlng

**Comment by miri64** [comment]
> Hm, doesn't look that *unintrusive* to me. A lot of modules are touched and I cannot test them all.

The unintrusive means that this PR doesn't change anything so that RIOT can't be run *without* `gnrc_netif2` anymore.

**Comment by miri64** [comment]
> What's missing (IMHO) is to also show in Makefiles of the adapted examples for instance which modules to use/exchange to enable netif2.

Since in this PR this is not linkable I think it's a bad idea to add documentation for that here.

**Comment by miri64** [comment]
> and I cannot test them all. 

I thought you were testing #7479 already.

**Comment by smlng** [comment]
> Since in this PR this is not linkable I think it's a bad idea to add documentation for that here.

but then I don't understand why you (need to) touch/adapt `examples/gnrc_minimal/main.c` and `examples/gnrc_border_router/Makefile`, without enabling netif2/ndp2 somewhere they won't use it anyway, right?

**Comment by miri64** [comment]
> but then I don't understand why you (need to) touch/adapt examples/gnrc_minimal/main.c and examples/gnrc_border_router/Makefile, without enabling netif2/ndp2 somewhere they won't use it anyway, right?

Because this PR touches everything that uses `gnrc_netif`, `gnrc_ipv6_netif`, `gnrc_sixlowpan_netif`, ... directly and replaces it with the according mechanism. If this is blocking we can revert the examples and only change them for the final switch-over to `gnrc_netif2` and NIB. But then you would have to hand-patch them every time you want to test the new NDP implementation. Right now you just have to `USEMODULE=gnrc_netif2` (in this PR) or `USEMODULE=gnrc_ipv6_nib` (or other submodules, in the other downstream PRs respectively).

**Comment by miri64** [comment]
TL;DR: To make testing easier for you guys.

**Comment by miri64** [comment]
I changed the base branch now to the freshly created `gnrc_netif2_integration/master` branch. I will remove all the `ifdef`s later.

**Comment by miri64** [comment]
> I will remove all the ifdefs later.

Seems like I already did that ^^" Ready for review.

File: Makefile.dep [hunk 2]
```diff
@@ -319,38 +313,6 @@ ifneq (,$(filter gnrc_ipv6_nc,$(USEMODULE)))
   USEMODULE += ipv6_addr
 endif
 
-ifneq (,$(filter gnrc_ipv6_nib_6lbr,$(USEMODULE)))
```
Comment (cgundogan) [inline Makefile.dep]: why do you remove the nib dependency information?

Comment (miri64) [reply Makefile.dep]: Probably rebase error.

Comment (miri64) [reply Makefile.dep]: Fixed

**Comment by miri64** [comment]
Mhhhhh problem with all the non-unintrusive porting now is: Murdock fails to build (actually only to link), since old NDP isn't ported (and I don't plan to). This *is* now merged into a seperate branch. Do we accept the temporary broken state?

**Comment by miri64** [comment]
Rebased to current gnrc_netif2_integration/master.

File: examples/gnrc_minimal/main.c [hunk 3]
```diff
@@ -22,25 +22,26 @@
...
-    size_t numof = gnrc_netif_get(ifs);
-    if (numof > 0) {
-        gnrc_ipv6_netif_t *entry = gnrc_ipv6_netif_get(ifs[0]);
-        for (int i = 0; i < GNRC_IPV6_NETIF_ADDR_NUMOF; i++) {
-            if ((ipv6_addr_is_link_local(&entry->addrs[i].addr)) && !(entry->addrs[i].flags & GNRC_IPV6_NETIF_ADDR_FLAGS_NON_UNICAST)) {
-                char ipv6_addr[IPV6_ADDR_MAX_STR_LEN];
-                ipv6_addr_to_str(ipv6_addr, &entry->addrs[i].addr, IPV6_ADDR_MAX_STR_LEN);
-                printf("My address is %s\n", ipv6_addr);
-            }
+    /* get interface and print their addresses */
```
[the remark this answers is not in the document]
Comment (miri64) [reply examples/gnrc_minimal/main.c]: "their" can be singular, but I guess this does not apply for things :stuck_out_tongue:

File: examples/gnrc_networking/Makefile [hunk 4]
```diff
@@ -22,7 +22,7 @@ USEMODULE += gnrc_ipv6_router_default
 USEMODULE += gnrc_udp
 # Add a routing protocol
 USEMODULE += gnrc_rpl
-USEMODULE += auto_init_gnrc_rpl
+# USEMODULE += auto_init_gnrc_rpl
```
[the remark this answers is not in the document]
Comment (miri64) [reply examples/gnrc_networking/Makefile]: Will remove.

File: examples/gnrc_minimal/main.c [hunk 5]
```diff
@@ -22,25 +22,26 @@
...
+    gnrc_netif2_t *netif = NULL;
+    while ((netif = gnrc_netif2_iter(netif))) {
+        ipv6_addr_t ipv6_addrs[GNRC_NETIF2_IPV6_ADDRS_NUMOF];
+        int res = gnrc_netapi_get(netif->pid, NETOPT_IPV6_ADDR, 0, ipv6_addrs,
+                                  sizeof(ipv6_addrs));
+
+        for (int i = 0; i < (res / sizeof(ipv6_addr_t)); i++) {
```
[the remark this answers is not in the document]
Comment (miri64) [reply examples/gnrc_minimal/main.c]: Will fix

**Comment by miri64** [comment]
@bergzand addressed your comments

File: sys/shell/commands/sc_fib.c [hunk 6]
```diff
@@ -208,10 +206,11 @@ int _fib_route_handler(int argc, char **argv)
     /* e.g. fibroute add <destination> via <next hop> lifetime <lifetime> */
     if ((argc == 7) && (strcmp("add", argv[1]) == 0) && (strcmp("via", argv[3]) == 0)
             && (strcmp("lifetime", argv[5]) == 0)) {
-        kernel_pid_t ifs[GNRC_NETIF_NUMOF];
-        size_t ifnum = gnrc_netif_get(ifs);
+                size_t ifnum = gnrc_netif2_numof();
```
Comment (bergzand) [inline sys/shell/commands/sc_fib.c]: Indentation looks off here

File: sys/shell/commands/sc_gnrc_6ctx.c [hunk 7]
```diff
@@ -72,6 +69,7 @@ static void _adv_ctx(void)
             gnrc_ndp_internal_send_rtr_adv(ifs[i], NULL, NULL, false);
         }
     }
+#endif
```
Comment (bergzand) [inline sys/shell/commands/sc_gnrc_6ctx.c]: This looks like a stray `#endif`

**Comment by bergzand** [comment]
Just to confirm, I'm getting compilation errors (not only linking errors) with the gnrc_networking example. All due to missing `gnrc_ipv6_nib_iface` functions:
```
/home/hydrazine/dev/RIOT/sys/net/gnrc/netif2/gnrc_netif2.c: In function ‘gnrc_netif2_set_from_netdev’:
/home/hydrazine/dev/RIOT/sys/net/gnrc/netif2/gnrc_netif2.c:281:21: error: implicit declaration of function ‘gnrc_ipv6_nib_iface_cease_rtr_adv’ [-Werror=implicit-function-declaration]
                     gnrc_ipv6_nib_iface_cease_rtr_adv(netif);
                     ^
/home/hydrazine/dev/RIOT/sys/net/gnrc/netif2/gnrc_netif2.c:290:17: error: implicit declaration of function ‘gnrc_ipv6_nib_iface_start_rtr_adv’ [-Werror=implicit-function-declaration]
                 gnrc_ipv6_nib_iface_start_rtr_adv(netif);
                 ^
/home/hydrazine/dev/RIOT/sys/net/gnrc/netif2/gnrc_netif2.c: In function ‘_gnrc_netif2_thread’:
/home/hydrazine/dev/RIOT/sys/net/gnrc/netif2/gnrc_netif2.c:1137:5: error: implicit declaration of function ‘gnrc_ipv6_nib_init_iface’ [-Werror=implicit-function-declaration]
     gnrc_ipv6_nib_init_iface(netif);
     ^
In file included from /home/hydrazine/dev/RIOT/sys/net/gnrc/network_layer/ipv6/nib/_nib-6ln.h:27:0,
                 from /home/hydrazine/dev/RIOT/sys/net/gnrc/network_layer/ipv6/nib/_nib-6ln.c:18:
/home/hydrazine/dev/RIOT/sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.h:50:38: error: unknown type name ‘gnrc_ipv6_netif_t’
 void _snd_ns(const ipv6_addr_t *tgt, gnrc_ipv6_netif_t *netif,
                                      ^
```

They are gone when testing with #7456, so I don't really mind as long as it is known.
gnrc_minimal compiles but fails to link as expected.

File: sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c [hunk 8]
```diff
@@ -129,8 +129,7 @@ gnrc_rpl_instance_t *gnrc_rpl_instance_get(uint8_t instance_id)
     return NULL;
 }
 
-bool gnrc_rpl_dodag_init(gnrc_rpl_instance_t *instance, ipv6_addr_t *dodag_id, kernel_pid_t iface,
-                         gnrc_ipv6_netif_addr_t *netif_addr)
+bool gnrc_rpl_dodag_init(gnrc_rpl_instance_t *instance, ipv6_addr_t *dodag_id, kernel_pid_t iface)
 {
     /* TODO: check if netif_addr belongs to iface */
```
[the remark this answers is not in the document]
Comment (miri64) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c]: I would like to keep it in for the person that is porting RPL (@cgundogan) so he can decide how they get the interface address.

**Comment by miri64** [comment]
> They are gone when testing with #7456, so I don't really mind as long as it is known.

It is known :-)

File: sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (bergzand) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c]: Ack, fine with me :)

**Comment by miri64** [comment]
Addressed comments

**Comment by bergzand** [comment]
Thanks for the changes and clarifications. Looks good to me now

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
@smlng can we dismiss your review?

**Comment by smlng** [comment]
@miri64 general question (and suggestion if you want): netif2 will replace netif, right? As you now PRed against a feature branch it would IMHO make perfect sense to use `netif` for the naming already. Otherwise this whole feature branch doesn't make sense at all, the idea would be get the full replacement working and tested within this branch and merge the feature at once, when everything is fine.

**Comment by miri64** [comment]
> @miri64 general question (and suggestion if you want): netif2 will replace netif, right? As you now PRed against a feature branch it would IMHO make perfect sense to use netif for the naming already. Otherwise this whole feature branch doesn't make sense at all, the idea would be get the full replacement working and tested within this branch and merge the feature at once, when everything is fine.

I will do that, but not in this PR. In fact that's intended to be the very last PR against #7925 before reintegrating it into master ;)

**Comment by miri64** [comment]
See also https://github.com/RIOT-OS/RIOT/projects/6#card-5336168 regarding that

**Comment by miri64** [comment]
Merged intentionally with broken Murdock state
